provider/datagen/src/transform/segmenter/mod.rs

```diff
@@ -703,6 +704,105 @@ impl IterableResourceProvider<SentenceBreakDataV1Marker> for SegmenterRuleProvid
     }
 }
 
+#[derive(serde::Deserialize, Debug)]
+struct SegmenterDictionaryData {
+    trie_data: Vec<u16>,
+}
+
+/// A data provider reading from segmenter dictionary files.
+#[derive(Debug)]
+pub struct SegmenterDictionaryProvider {
+    source: SourceData,
+}
+
+impl SegmenterDictionaryProvider {
+    fn get_toml_filename(options: &ResourceOptions) -> Result<&'static str, DataError> {
+        if options.get_langid() == langid!("km") {
+            Ok("dictionary_km.toml")
+        } else if options.get_langid() == langid!("ja") {
+            Ok("dictionary_cj.toml")
+        } else if options.get_langid() == langid!("lo") {
+            Ok("dictionary_lo.toml")
+        } else if options.get_langid() == langid!("my") {
+            Ok("dictionary_my.toml")
+        } else if options.get_langid() == langid!("th") {
+            Ok("dictionary_th.toml")
+        } else {
+            Err(DataError::custom("Unsupported options"))
+        }
+    }
+}
+
+impl From<&SourceData> for SegmenterDictionaryProvider {
+    fn from(source: &SourceData) -> Self {
+        Self {
+            source: source.clone(),
+        }
+    }
+}
+
+impl ResourceProvider<UCharDictionaryBreakDataV1Marker> for SegmenterDictionaryProvider {
+    fn load_resource(
+        &self,
+        req: &DataRequest,
+    ) -> Result<DataResponse<UCharDictionaryBreakDataV1Marker>, DataError> {
+        if req.options.get_langid() == langid!("ja")
+            || req.options.get_langid() == langid!("km")
+            || req.options.get_langid() == langid!("lo")
+            || req.options.get_langid() == langid!("my")
+            || req.options.get_langid() == langid!("th")
+        {
+            let toml_data = self
+                .source
+                .segmenter()?
+                .read_and_parse_toml::<SegmenterDictionaryData>(Self::get_toml_filename(
+                    &req.options,
+                )?)?;
```

Comment:
Suggestion: make `get_toml_filename` return `Option<&'static str>`, and then do

```suggestion
                .read_and_parse_toml::<SegmenterDictionaryData>(Self::get_toml_filename(
                    &req.options,
                ).ok_or_else(|| DataErrorKind::MissingResourceOptions.with_req(UCharDictionaryBreakDataV1Marker::KEY, req))?)?;
```

and then you can remove the redundant `if req.options.get_langid() == ...` condition in this function.

Note that you should be returning `DataErrorKind::MissingResourceOptions`, not `DataError::custom`